Add processor parse container log native

quzard opened the pull request.

onejimmyboy committed:
Update input-file.md (#1353)

quzard committed:
ContainerdLogLineParser

quzard committed:
add DockerJsonLogLineParser

quzard committed:
add AddDockerJsonLog

quzard committed:
Merge branch 'add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
RegisterProcessorCreator

quzard committed:
去除docker末尾的回车

quzard committed:
modify TestDockerJsonLogLineParserWithSplit

quzard committed:
containerd-text

quzard committed:
containerd-text

quzard committed:
containerd-text

quzard committed:
FILE_ENCODING

quzard committed:
Merge remote-tracking branch 'origin/add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
参数放到常量里

quzard committed:
Merge branch 'add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
Merge branch 'add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
Merge branch 'add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
MergeType

henryzhx8 commented on core/models/PipelineEventGroup.h:
```diff
@@ -41,6 +41,7 @@ enum class EventGroupMetaKey {
     LOG_FILE_INODE,
     LOG_READ_OFFSET,
     LOG_READ_LENGTH,
+    FILE_ENCODING,
```
这是啥？

quzard replied:
用于Meta data

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    const StringView& containerType = logGroup.GetMetadata(EventGroupMetaKey::FILE_ENCODING);
+    EventsContainer& events = logGroup.MutableEvents();
+
+    for (auto it = events.begin(); it != events.end();) {
+        if (ProcessEvent(containerType, *it)) {
+            ++it;
+        } else {
+            it = events.erase(it);
+        }
+    }
```
特别注意有没有迭代器失效问题

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    }
+}
+
+bool ProcessorParseContainerLogNative::ProcessEvent(const StringView& containerType, PipelineEventPtr& e) {
```
所有的stringview都直接按值传递

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    if (!parseSuccess) {
+        return true;
+    }
+    if (!doc.HasMember(DOCKER_JSON_LOG.c_str()) || !doc.HasMember(DOCKER_JSON_TIME.c_str())
```
不要先hasmember，再getmember，直接findmember，然后用迭代器，效率会高很多

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    return true;
+}
+
+void ProcessorParseContainerLogNative::AddDockerJsonLog(char ** data ,const StringView& key, const StringView& value, LogEvent& targetEvent) {
+    memmove(*data, key.data(), key.size());
+    StringView keyBuffer = StringView(*data, key.size());
+    *data += key.size();
+    memmove(*data, value.data(), value.size());
+    StringView valueBuffer = StringView(*data, value.size());
+    *data += value.size();
+    AddLog(keyBuffer, valueBuffer, targetEvent);
+}
```
这个有点hack了，虽然看起来是对的，但感觉风险很大，越界检查什么都没有。既然这样子弄了，那为啥不更直接点，在原始日志上硬解，还能更省内存。

quzard replied:
在原始日志上硬解也要分配临时空间存储json解析后的content。自己要实现一个content的json解析函数

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -0,0 +1,57 @@
+    bool IsSupportedEvent(const PipelineEventPtr& e) const override;
+
+private:
+    std::string containerTimeKey = "_time_"; // 容器时间字段
+    std::string containerSourceKey = "_source_"; // 容器来源字段
+    std::string containerLogKey = "content"; // 容器日志字段
```
这个插件是一个完全新的插件还是替代go插件？如果是新的话，这些key的名字可以重新斟酌一下，包括time这个字段是不是直接不要了，默认作为__time__

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    AddLog(keyBuffer, valueBuffer, targetEvent);
+}
+
+void ProcessorParseContainerLogNative::AddLog(const StringView& key,
```
这个包一层的意义？

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    }
+}
+
+bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceEvent, PipelineEventPtr& e) {
```
这个e干啥用了？

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+}
+
+bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceEvent, PipelineEventPtr& e) {
+    StringView buffer = sourceEvent.GetContent(mSourceKey);
```
源字段删除考虑了吗？虽然恰好前后都是content，但不要将这种条件隐式表达。

quzard replied:
如果源字段被删除了，GetContent会返回一个空的StringView。下面有判断空的逻辑

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+bool ProcessorParseContainerLogNative::ContainerdLogLineParser(LogEvent& sourceEvent, PipelineEventPtr& e) {
+    StringView contentValue = sourceEvent.GetContent(mSourceKey);
+
+    if (contentValue.empty() || contentValue.length() < 5)
```
小于5是什么意思？

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    // 寻找第一个分隔符位置 时间 _time_
+    StringView timeValue;
+    const char* pch1
+        = std::search(contentValue.begin(), contentValue.end(), CONTIANERD_DELIMITER.begin(), CONTIANERD_DELIMITER.end());
```
这个函数的效率怎么样？和find比

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    }
+    sourceValue = StringView(pch1 + 1, pch2 - pch1 - 1);
+
+    if (sourceValue != "stdout" && sourceValue != "stderr")
```
docker那为什么没有这个？

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+
+    // 如果既不以 CONTIANERD_PART_TAG 开头，也不以 CONTIANERD_FULL_TAG 开头
+    if (*(pch2 + 1) != CONTIANERD_PART_TAG && *(pch2 + 1) != CONTIANERD_FULL_TAG) {
+        StringBuffer containerTimeKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerTimeKey);
+        StringBuffer containerSourceKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerSourceKey);
+        StringBuffer containerLogKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerLogKey);
```
这几个一样的可以一个logGroup共享一个吗？

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+    StringView timeValue;
+    const char* pch1
+        = std::search(contentValue.begin(), contentValue.end(), CONTIANERD_DELIMITER.begin(), CONTIANERD_DELIMITER.end());
+    if (pch1 >= contentValue.end() - 1) {
```
这里的条件为什么和下面两个不一样？

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,293 @@
+        StringView content = StringView(pch2 + 1, contentValue.end() - pch2 - 1);
+        AddLog(StringView(containerTimeKeyBuffer.data, containerTimeKeyBuffer.size), timeValue, sourceEvent);
+        AddLog(StringView(containerSourceKeyBuffer.data, containerSourceKeyBuffer.size), sourceValue, sourceEvent);
+        AddLog(StringView(containerLogKeyBuffer.data, containerLogKeyBuffer.size), content, sourceEvent);
+        return true;
+    }
+
+    // 寻找第三个分隔符位置
+    const char* pch3
+        = std::search(pch2 + 1, contentValue.end(), CONTIANERD_DELIMITER.begin(), CONTIANERD_DELIMITER.end());
+    if (pch3 == contentValue.end() || pch3 != pch2 + 2) {
+        return true;
+    }
+    // F
+    if (*(pch2 + 1) == CONTIANERD_FULL_TAG) {
+        StringBuffer containerTimeKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerTimeKey);
+        StringBuffer containerSourceKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerSourceKey);
+        StringBuffer containerLogKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerLogKey);
+        // content
+        StringView content = StringView(pch3 + 1, contentValue.end() - pch3 - 1);
+        AddLog(StringView(containerTimeKeyBuffer.data, containerTimeKeyBuffer.size), timeValue, sourceEvent);
+        AddLog(StringView(containerSourceKeyBuffer.data, containerSourceKeyBuffer.size), sourceValue, sourceEvent);
+        AddLog(StringView(containerLogKeyBuffer.data, containerLogKeyBuffer.size), content, sourceEvent);
+        return true;
+    }
+    // P
+    if (*(pch2 + 1) == CONTIANERD_PART_TAG) {
+        StringBuffer containerTimeKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerTimeKey);
+        StringBuffer containerSourceKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerSourceKey);
+        StringBuffer containerLogKeyBuffer = sourceEvent.GetSourceBuffer()->CopyString(containerLogKey);
+        StringBuffer partTagBuffer = sourceEvent.GetSourceBuffer()->CopyString(PARTLOGFLAG);
+        // content
+        StringView content = StringView(pch3 + 1, contentValue.end() - pch3 - 1);
+        StringView partTag = StringView(pch2 + 1, 1);
+        AddLog(StringView(containerTimeKeyBuffer.data, containerTimeKeyBuffer.size), timeValue, sourceEvent);
+        AddLog(StringView(containerSourceKeyBuffer.data, containerSourceKeyBuffer.size), sourceValue, sourceEvent);
+        AddLog(StringView(partTagBuffer.data, partTagBuffer.size), partTag, sourceEvent);
+        AddLog(StringView(containerLogKeyBuffer.data, containerLogKeyBuffer.size), content, sourceEvent);
+        return true;
+    }
```
这一堆适当合并下吧，太啰嗦了

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
这里有没有考虑常规格式之外突然出现一下异常信息的情况？

quzard replied:
docker_json-file靠是否json解析成功来判断。containerd_text 有判断日志格式是否合法。不合法就直接返回原始日志

quzard committed:
const StringView

quzard committed:
优化json find逻辑

quzard committed:
Merge branch 'add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
modify AddDockerJsonLog

quzard committed:
std::find

quzard committed:
AddContainerdTextLog

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -100,33 +118,32 @@ bool ProcessorParseContainerLogNative::ProcessEvent(const StringView& containerT
 
-    if (contentValue.empty() || contentValue.length() < 5)
+    if (contentValue.empty())
         return true;
```
这个判断条件省了吧，其它解析插件也没单独拎出来

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -100,33 +118,32 @@ bool ProcessorParseContainerLogNative::ProcessEvent(const StringView& containerT
-        = std::search(contentValue.begin(), contentValue.end(), CONTIANERD_DELIMITER.begin(), CONTIANERD_DELIMITER.end());
-    if (pch1 >= contentValue.end() - 1) {
+    const char* pch1 = std::find(contentValue.begin(), contentValue.end(), CONTIANERD_DELIMITER);
+    if (pch1 >= contentValue.end()) {
```
```suggestion
    if (pch1 == contentValue.end()) {
```

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -142,53 +159,37 @@ bool ProcessorParseContainerLogNative::ContainerdLogLineParser(LogEvent& sourceE
-        AddLog(StringView(containerLogKeyBuffer.data, containerLogKeyBuffer.size), content, sourceEvent);
+        AddContainerdTextLog(keyBuffer, timeValue, sourceValue, content, partTag, sourceEvent);
         return true;
     }
```
这里有明显的逻辑漏洞，以下几种情况都会有问题：
1. 时间 stdout F（消息就是F）
2. 时间 stdout Fa msg（消息是Fa msg）

正确的逻辑分解方法：
1. 首先从pch2+1开始找第三个分隔符，如果找不到，说明pch2+1到结束的都是content（消息就是F这种）
2. 如果找到第三个分隔符pch3，则判断pch3-pch2是否等于2且pch2+1是否是F或者P，如果符合，说明pch3+1到结束是content（常规）
4. 否则，pch2+1到结束都是content（消息就是Fa msg这种）

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -100,33 +118,32 @@ bool ProcessorParseContainerLogNative::ProcessEvent(const StringView& containerT
+    const char* pch1 = std::find(contentValue.begin(), contentValue.end(), CONTIANERD_DELIMITER);
+    if (pch1 >= contentValue.end()) {
         // 没有找到分隔符
         return true;
```
像这种地方，其实都是异常情况，需要打warning日志的，所以函数签名里还要加一个errorMsg的出参，到外层的ProcessEvent函数里去打印具体的日志。

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -240,50 +247,84 @@ bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceE
+                                                            StringView timeValue,
+                                                            StringView sourceValue,
+                                                            StringView content,
+                                                            StringView partTag,
```
这个参数直接用bool isPartialLog

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -0,0 +1,82 @@
+private:
+    std::string containerTimeKey = "_time_"; // 容器时间字段
+    std::string containerSourceKey = "_source_"; // 容器来源字段
+    std::string containerLogKey = "content"; // 容器日志字段
```
这个也改成static const

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -0,0 +1,82 @@
+public:
+    static const std::string sName;
+
+    static const char CONTIANERD_DELIMITER; // 分隔符
+    static const char CONTIANERD_FULL_TAG; // 容器全标签
+    static const char CONTIANERD_PART_TAG; // 容器部分标签
+
+    static const std::string DOCKER_JSON_LOG; // docker json 日志字段
+    static const std::string DOCKER_JSON_TIME; // docker json 时间字段
+    static const std::string DOCKER_JSON_STREAM_TYPE; // docker json 流字段
```
这些其它类会用吗？不会的话放private

quzard replied:
回退逻辑里需要

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -0,0 +1,82 @@
+
+namespace logtail {
+
+struct ContainerStdoutKeyStringBuffer {
+    StringBuffer timeKeyBuffer;
+    StringBuffer sourceKeyBuffer;
+    StringBuffer logKeyBuffer;
+    StringBuffer flagBuffer;
+};
```
这个不需要了，containerTimeKey这几个参数变成类常量以后，setcontent的时候可以直接stringview引用，不需要再CopyString了，因为相当于全局变量，生命周期肯定比loggroup长

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -44,11 +59,21 @@ class ProcessorParseContainerLogNative : public Processor {
+                      const StringView containerType,
+                      PipelineEventPtr& e,
+                      ContainerStdoutKeyStringBuffer* containerStdoutKeyStringBuffer);
+    void AddDockerJsonLog(char** data, const StringBuffer key, const StringView value, LogEvent& targetEvent);
```
1. StringView不需要const修饰
2. char**不太常用，还是改成两个参数吧，加一个index记录当前位置

根据之前的描述，可以直接改成
```suggestion
    void AddDockerJsonLog(char* data, size_t &index, StringView key, StringView value, LogEvent& targetEvent);
```

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,334 @@
+                                                               ContainerStdoutKeyStringBuffer* keyBuffer) {
+    StringView buffer = sourceEvent.GetContent(mSourceKey);
+
+    if (buffer.empty())
+        return true;
```
```suggestion
```

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -0,0 +1,334 @@
+    bool parseSuccess = true;
+    rapidjson::Document doc;
+    doc.Parse(buffer.data(), buffer.size());
+    if (doc.HasParseError()) {
+        if (LogtailAlarm::GetInstance()->IsLowLevelAlarmValid()) {
+            LOG_WARNING(sLogger,
+                        ("parse docker stdout json log fail, log",
+                         buffer)("rapidjson offset", doc.GetErrorOffset())("rapidjson error", doc.GetParseError())(
+                            "project", GetContext().GetProjectName())("logstore", GetContext().GetLogstoreName()));
+            LogtailAlarm::GetInstance()->SendAlarm(PARSE_LOG_FAIL_ALARM,
+                                                   std::string("parse json fail:") + buffer.to_string(),
+                                                   GetContext().GetProjectName(),
+                                                   GetContext().GetLogstoreName(),
+                                                   GetContext().GetRegion());
+        }
+        parseSuccess = false;
+    } else if (!doc.IsObject()) {
+        if (LogtailAlarm::GetInstance()->IsLowLevelAlarmValid()) {
+            LOG_WARNING(sLogger,
+                        ("invalid docker stdout json object, log",
+                         buffer)("project", GetContext().GetProjectName())("logstore", GetContext().GetLogstoreName()));
+            LogtailAlarm::GetInstance()->SendAlarm(PARSE_LOG_FAIL_ALARM,
+                                                   std::string("invalid json object:") + buffer.to_string(),
+                                                   GetContext().GetProjectName(),
+                                                   GetContext().GetLogstoreName(),
+                                                   GetContext().GetRegion());
+        }
+        parseSuccess = false;
+    }
+    if (!parseSuccess) {
+        return true;
+    }
```
和containerdParse一样，解析类错误通过errorMsg传递出去，在ProcessEvent里统一处理

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -240,50 +247,84 @@ bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceE
-    StringView timeValue(doc[DOCKER_JSON_TIME.c_str()].GetString());
+    }
+    StringView timeValue = findMemberAndGetString(DOCKER_JSON_TIME);
+    if (sourceValue.empty()) {
+        if (LogtailAlarm::GetInstance()->IsLowLevelAlarmValid()) {
+            LOG_WARNING(sLogger,
+                        ("invalid docker stdout json log, log",
+                         buffer)("project", GetContext().GetProjectName())("logstore", GetContext().GetLogstoreName()));
+            LogtailAlarm::GetInstance()->SendAlarm(PARSE_LOG_FAIL_ALARM,
+                                                   std::string("invalid docker stdout json log:") + buffer.to_string(),
+                                                   GetContext().GetProjectName(),
+                                                   GetContext().GetLogstoreName(),
+                                                   GetContext().GetRegion());
+        }
+        return true;
+    }
```
同上，类似的错误处理统一由外部处理

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -226,8 +227,14 @@ bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceE
-    if (!doc.HasMember(DOCKER_JSON_LOG.c_str()) || !doc.HasMember(DOCKER_JSON_TIME.c_str())
-        || !doc.HasMember(DOCKER_JSON_STREAM_TYPE.c_str())) {
+
+    auto findMemberAndGetString = [&](const std::string& memberName) {
+        auto it = doc.FindMember(memberName.c_str());
+        return it != doc.MemberEnd() ? StringView(it->value.GetString()) : StringView();
+    };
```
这个放到JsonUtil里吧

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -226,8 +227,14 @@ bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceE
+
+    auto findMemberAndGetString = [&](const std::string& memberName) {
+        auto it = doc.FindMember(memberName.c_str());
+        return it != doc.MemberEnd() ? StringView(it->value.GetString()) : StringView();
```
```suggestion
        return it != doc.MemberEnd() ? StringView(it->value.GetCString()) : StringView();
```
可以省一次string构造

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.cpp:
```diff
@@ -240,50 +247,84 @@ bool ProcessorParseContainerLogNative::DockerJsonLogLineParser(LogEvent& sourceE
+    if (buffer.size() < content.size() + timeValue.size() + sourceValue.size()) {
+        StringBuffer stringBuffer = logGroup.GetSourceBuffer()->AllocateStringBuffer(content.size() + timeValue.size()
+                                                                                     + sourceValue.size() + 1);
+        data = const_cast<char*>(stringBuffer.data);
```
这个也应该属于非预期情况，需要warning，且直接返回，不要再继续了

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -44,11 +59,21 @@ class ProcessorParseContainerLogNative : public Processor {
+                      PipelineEventPtr& e,
+                      ContainerStdoutKeyStringBuffer* containerStdoutKeyStringBuffer);
+    void AddDockerJsonLog(char** data, const StringBuffer key, const StringView value, LogEvent& targetEvent);
+    void AddContainerdTextLog(ContainerStdoutKeyStringBuffer* keyBuffer,
+                 StringView timeValue,
+                 StringView sourceValue,
+                 StringView content,
+                 StringView partTag,
+                 LogEvent& sourceEvent);
+    bool ContainerdLogLineParser(LogEvent& sourceEvent, ContainerStdoutKeyStringBuffer* containerStdoutKeyStringBuffer);
+    bool DockerJsonLogLineParser(PipelineEventGroup& logGroup,
+                                 LogEvent& sourceEvent,
+                                 ContainerStdoutKeyStringBuffer* containerStdoutKeyStringBuffer);
```
综合之前的所有描述，函数签名改成如下：
```suggestion
    void AddContainerdTextLog(StringView time, StringView source, StringView content, bool isPartialLog, LogEvent& sourceEvent);
    bool ParseContainerdLogLine(LogEvent& sourceEvent, std::string& errorMsg);
    bool ParseDockerJsonLogLine(LogEvent& sourceEvent, std::string& errorMsg);
```

henryzhx8 commented on core/processor/ProcessorParseContainerLogNative.h:
```diff
@@ -44,11 +59,21 @@ class ProcessorParseContainerLogNative : public Processor {
-    void AddLog(const StringView& key, const StringView& value, LogEvent& targetEvent, bool overwritten = true);
-    bool ContainerdLogLineParser(LogEvent& sourceEvent, PipelineEventPtr& e);
-    bool DockerJsonLogLineParser(LogEvent& sourceEvent, PipelineEventPtr& e);
+    bool ProcessEvent(PipelineEventGroup& logGroup,
+                      const StringView containerType,
+                      PipelineEventPtr& e,
+                      ContainerStdoutKeyStringBuffer* containerStdoutKeyStringBuffer);
```
```suggestion
    bool ProcessEvent(StringView containerType, PipelineEventPtr& e);
```

yyuuttaaoo committed:
Fix: Ensure process is able to exit when history file import is ever used (#1369)

henryzhx8 committed:
Fix: Resolve config loading and provider destruct order issues (#1360)

1. fix pluging registration for golang plugin to recognize enterprise golang plugin
2. stop ConfigProvider explicitly instead of letting it end itself during class destruction to avoid rare core dump during program termination
3. load old go pipeline when existing config becomes invalid, so that the golang pipeline remains running after pipeline resumes.

quzard committed:
Merge branch 'add-ProcessorMergeMultilineLogNative' into add-ProcessorParseContainerLogNative

quzard committed:
modify

quzard committed:
modify

Takuka0311 committed:
self-hosted github action runners (#1371)

quzard committed:
Merge remote-tracking branch 'alibaba/main' into add-ProcessorParseContainerLogNative
